src/clr-angular/utils/focus/key-focus/key-focus.ts

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { EventEmitter, HostListener, Input, Output, Component, ContentChildren, QueryList } from '@angular/core';
+import { Subscription } from 'rxjs';
+
+import { ClrKeyFocusItem } from './key-focus-item';
+import { ClrFocusDirection } from './enums/focus-direction.enum';
+import { KeyCodes, IEKeyCodes } from './../key-codes.enum';
+import { FocusableItem } from './interfaces';
+import { preventArrowKeyScroll } from './util';
+
+@Component({
+  selector: '[clrKeyFocus]',
+  template: '<ng-content></ng-content>',
+})
+export class ClrKeyFocus {
+  @Input('clrDirection') direction = ClrFocusDirection.VERTICAL;
+  @Input('clrFocusOnLoad') focusOnLoad = false;
+  @Output('clrFocusChange') private focusChange: EventEmitter<void> = new EventEmitter<void>();
+  @ContentChildren(ClrKeyFocusItem, { descendants: true })
+  private clrKeyFocusItems: QueryList<ClrKeyFocusItem>;
+
+  private _focusableItems: Array<FocusableItem>;
+  @Input('clrKeyFocus')
+  set focusableItems(elements: Array<FocusableItem>) {
+    // We accept a list of focusable elements (HTMLElements or existing Directives) or auto query for clrKeyFocusItem
+    // We accept a list reference in the cases where we cannot use ContentChildren to query
+    // ContentChildren can be unavailable if content is projected outside the scope of the component (see tabs).
+    if (elements && elements.length) {
+      this._focusableItems = elements;
+      this.initializeFocus();
+    }
+  }
+
+  get focusableItems() {
+    if (this._focusableItems) {
+      return this._focusableItems;
+    } else {
+      return this.clrKeyFocusItems.toArray();
+    }
+  }
+
+  private _current: number = 0;
+  get current() {
+    return this._current;
+  }
+
+  private subscriptions: Subscription[] = [];
+
+  ngAfterContentInit() {
+    this.subscriptions.push(this.listenForItemUpdates());
+    this.initializeFocus();
+  }
+
+  ngOnDestroy() {
+    this.subscriptions.forEach(s => s.unsubscribe());
+  }
+
+  @HostListener('keydown', ['$event'])
+  handleKeyboardEvent(event: KeyboardEvent) {
+    if (this.prevKeyPressed(event) && this.currentFocusIsNotFirstItem()) {
+      this.keyAction(() => this._current--);
+    } else if (this.nextKeyPressed(event) && this.currentFocusIsNotLastItem()) {
+      this.keyAction(() => this._current++);
+    } else if (event.code === KeyCodes.Home) {
+      this.keyAction(() => (this._current = 0));
+    } else if (event.code === KeyCodes.End) {
+      this.keyAction(() => (this._current = this.focusableItems.length - 1));
+    }
+
+    preventArrowKeyScroll(event);
+  }
+
+  @HostListener('click', ['$event'])
+  handleClickEvent(event: any) {
+    let position;
+
+    if (this.focusableItems[0].nativeElement) {
```

Comment:
I am assuming this is here to check if the event is sourced from a focusKeyItem's element. But I think this line is not guaranteed to find the right focusKeyItem. For example, if a user placed `clrFocusKeyItem` on a component or a parent element, and the event is originated from the inside nested child element, focusKeyItem .nativeElement wouldn't match that child element.

Reply:
That would mean they have nested clickable targets, which we shouldn't be supporting. You're right, the `event.target` would be wrong in that case, but I can't think of a case where we'd support this. The rest of the function will just go without firing anything since the `else` block will not find the `event.target` in the focusable items, which seems like a reasonable code path in that case.

Reply:
Right this is not designed to handle nested click targets, if you need nested click targets then the consumer of the directive should get the proper references and pass them into the `clrKeyFocus`. This is essentially what the tabs are doing. `FocusableItems` can be an Angular Element ref or a native DOM element (Either meeting the `FocusableItem` interface).